### DISCUSSION TITLE
Getting involved: how to (kotlin edition)?

### DISCUSSION THREAD

**Discussion by fpoyer** [openrewrite/rewrite#3173] [Q&A] [answered]
Hi all,
I discovered Open rewrite a few days ago and I'm in love with the idea!
Keeping dependencies up to date or being able to apply large scale refactorings while using a semantic level logic is what I've been looking for for a long time without knowing it.

I am currently working an a huge code base (several thousands of mixed java and kotlin files) with some pretty old legacy code (dating back from the early 2000's) and I'm trying to get rid of PowerMock by using the dedicated recipe (org.openrewrite.java.testing.mockito.ReplacePowerMockito), but sadly it fails on some points at the moment:
- some (but not all) kotlin files are not parsed (and thus not affected), with the following message: "There were problems parsing XXX sources, run with --info to see full stack traces:" followed by list of ~50% of my kotlin files, and finally "Execution will continue but these files are unlikely to be affected by refactoring recipes."
- "when" is a keyword in Kotlin, so calls to static method PowerMockito#when are actually written as ``PowerMockito.`when` `` (with the back ticks) in Kotlin, and thus not detected (and not refactored) by the (java?) parser
- when removing the PowerMockIgnore annotation from (kotlin) test classes, sometimes a whole lot of the class is removed as well, hinting of incorrect detection of the annotation

Rather than complaining or requesting solutions be provided, I'd like to get involved as I feel like this tool could be immensely useful to me in the future, and should be more widely used everywhere.
But here is my problem now: I'm not sure how and where to start.
1. I feel like I should be adding some tests to replicate my errors, but I don't know if it should be in the rewrite-kotlin repo or in the rewrite-testing-framework that provides the recipe (or somewhere else)
2. Tests (at least in the rewrite-kotlin repo) are using a DSL that I'm having a hard time understanding what it does and how to use it. Is there any resource  (doc, forum thread, medium article, anything really) that can help me write a first test such as "with initial file X, when applying recipe Y, expected result is Z" or something like that?
Thanks for your help!

**Comment by timtebeek** [comment]
Hi @fpoyer ; thank you for the offer to help! I'm going to tag in @traceyyoshima here, who works on our Kotlin implementation, such that he can correct me when I'm wrong. ;)

It sounds like you're mostly looking to tweak the PowerMock recipe, such that it'll work well (or at least better) with Kotlin. The tests there will likely go into [rewrite-testing-frameworks](https://github.com/openrewrite/rewrite-testing-frameworks). Whereas the language level fixes will likely go into [rewrite-kotlin](https://github.com/openrewrite/rewrite-kotlin). You can work on these in tandem through `./gradlew publishToMavenLocal` on rewrite-kotlin, which should then be picked up in rewrite-testing-framework.

For the tests you'll likely want to add a test scoped dependency on rewrite-kotlin, [similar to what we have for rewrite-groovy already](https://github.com/openrewrite/rewrite-testing-frameworks/blob/c83feb434f28a28db1068be1105ac5cde03fbc09/build.gradle.kts#L42). Here's an [example of a Groovy test](https://github.com/openrewrite/rewrite-testing-frameworks/blob/c83feb434f28a28db1068be1105ac5cde03fbc09/src/test/java/org/openrewrite/java/testing/junit5/TemporaryFolderToTempDirTest.java#L49) of a [the TemporaryFolderToTempDir recipe](https://github.com/openrewrite/rewrite-testing-frameworks/blob/main/src/main/java/org/openrewrite/java/testing/junit5/TemporaryFolderToTempDir.java).
In your case you'll want to add a similar test using `org.openrewrite.kotlin.Assertions.kotlin` inside either [PowerMockitoMockStaticToMockitoTest](https://github.com/openrewrite/rewrite-testing-frameworks/blob/main/src/test/java/org/openrewrite/java/testing/mockito/PowerMockitoMockStaticToMockitoTest.java) or [ReplacePowerMockitoIntegrationTest](https://github.com/openrewrite/rewrite-testing-frameworks/blob/main/src/test/java/org/openrewrite/java/testing/mockito/ReplacePowerMockitoIntegrationTest.java).

The language level fixes I'm not as familiar with myself yet, but Tracey can likely help you out there if needed. I suppose your project is not open source, but you're welcome to report individual issues, or ideally a PR containing a unit test using the `Assertions.kotlin` seen above. That way they are easier to pick up on our end immediately.

We have [documented the test structure](https://docs.openrewrite.org/authoring-recipes/recipe-testing#writing-tests) as best we can; the docs assume Java, but with the import above you can just as easily test groovy, kotlin, or any other language we support.

Feel free to ask any follow up questions here, or [on our public Slack](https://join.slack.com/t/rewriteoss/shared_invite/zt-nj42n3ea-b~62rIHzb3Vo0E1APKCXEA); or even in the [OpenRewrite office hours](https://twitter.com/OpenRewrite/status/1651647628138913795) I'll be hosting at 16:00 CET today.

**Reply by timtebeek** [reply, marked as answer]
Cross referencing [the issue](https://github.com/openrewrite/rewrite-kotlin/pull/105) and [initial pull request](https://github.com/openrewrite/rewrite-kotlin/pull/106) here for visibility.
